### DISCUSSION TITLE
Derived atoms of ✨potentially asynchronous✨ atoms.

### DISCUSSION THREAD

**Discussion by iwoplaza** [pmndrs/jotai#2640] [Ideas] [answered]
### The problem

Atoms for query/data management libraries (e.g. `jotai-apollo`) usually have a signature that looks something like this:
```ts
const userAtom: Atom<User | Promise<User>> = ...;
```

If no data is cached yet, the atom holds a Promise, but if the data is already there and an optimistic update was performed, the atom's data is updated in sync and the atom can hold a concrete value instead of a Promise. This dual nature is nice (concrete value or a promise) because derived atoms could process the value in sync if it is available, and fallback to awaiting and returning a Promise if it is not.

However, handling those cases looks to be more trouble that it is worth, so the easiest is to just **always await**.

```ts
const userAtom: Atom<User | Promise<User>> = ...;

// typeof uppercaseNameAtom is Atom<Promise<String>>
const uppercaseNameAtom = atom(async (get) => {
  const user = await get(userAtom);
  return user.name.toUppercase();
});
```

That is what I have done in my code as well, but because of that, the parts of the app that use any atom that derives `userAtom` will suspend on synchronous updates to the `userAtom`, causing tiny loading states and unnecessary remounts. I recently had an idea for a better way, lets try to build up to it gradually.

### A potential solution

Below is an atom that handles both cases separately (processes the data in sync if it is available, returns a promise otherwise).

```ts
const userAtom: Atom<User | Promise<User>> = ...;

// typeof uppercaseNameAtom is Atom<string | Promise<String>>, which means we can derive it
// again and get the same dual behavior.
const uppercaseNameAtom = atom((get) => {
  const process = (user: User) => user.name.toUppercase();

  const userOrPromise = get(userAtom);
  if (isPromise(userOrPromise)) {
    // data not available yet, process it asynchronously
    return userOrPromise.then(process);
  }

  // data is here, process it now
  return process(userOrPromise);
});
```

It can be a bit too verbose though, and tougher to do more complex processing with. Let's try to simplify this a little. Say we have a function `soon` with the following signature:
```ts
type PromiseOrValue<T> = T | Promise<T>;

function soon<T, R>(data: PromiseOrValue<T>, process: (data: T) => R): PromiseOrValue<R> {
  // run `process` synchronously if `data` is known, return promise otherwise.
}
```

```ts
const userAtom: Atom<User | Promise<User>> = ...;

// typeof uppercaseNameAtom is Atom<string | Promise<String>>
const uppercaseNameAtom = atom(
  (get) => soon(get(userAtom), (user) => user.name.toUppercase())
);
```

For multiple dependencies, I made a `Promise.all` equivalent aptly called `soonAll`, which is an identity function if all arguments are known, and returns a Promise to an array with the concrete input values otherwise.

Since most of my derived atoms were just a few sync/async dependencies that were transformed in some way, I wanted to create a simple helper to reduce the amount of boilerplate I had to write:

```ts
const userAtom: Atom<User | Promise<User>> = ...;

// typeof uppercaseNameAtom is Atom<string | Promise<String>>
const uppercaseNameAtom = derive([userAtom], (user) => user.name.toUppercase());
```

Or an example with multiple dependencies:

```ts
const usersAtom: Atom<User[] | Promise<User[]>> = ...;
const nameFilterAtom = atom('');

// typeof filteredUsersAtom is Atom<User[] | Promise<User[]>>
const filteredUsersAtom = derive(
  [usersAtom, nameFilterAtom],
  (users, nameFilter) => {
    if (!nameFilter) {
      return users;
    }
    
    return users.filter((user) => user.name.includes(nameFilter));
  }
);
```

### Potential enhancements
So far I explored making the `soon` inspect the passed in Promise for a `status` field, and if it is `fulfilled`, it returns the value of the Promise. I wrote a couple tests for it, and it seems to be working fine. The main advantage of this approach, is that we can `derive` fully async atoms, and still be able to process the data in sync (for example, we change a sync dependency, but we also depend on 3 async atoms that are already fulfilled).

### Conclusion
This has thus far been a real help in my codebase, decreasing redundant computations. If this is too specific to be included in the library, I would be happy to create a separate module, maybe `jotai-derive`?

**Comment by dai-shi** [comment, marked as answer]
Thanks for opening up a discussion.

> This dual nature is nice (concrete value or a promise) because derived atoms could process the value in sync if it is available, and fallback to awaiting and returning a Promise if it is not.

Yes, this is pretty interesting. I knew this was possible and thought it's kind of advanced usage.



> However, handling those cases looks to be more trouble that it is worth, so the easiest is to just **always await**.

Exactly.

> causing tiny loading states and unnecessary remounts.

Exactly.

> Below is an atom that handles both cases separately (processes the data in sync if it is available, returns a promise otherwise).

I'm so glad to see you've reached there. I think I did something similar somewhere.





> Say we have a function `soon` with the following signature:

Niceeee.



> If this is too specific to be included in the library, I would be happy to create a separate module, maybe `jotai-derive`?

I'm very excited to see your solution.
Yes, my suggestion is to create a separate module. (Mainly due to that it allows you rapid development without constraints. We might consider adding the outcome in this library, but my gut feeling is that it's not necessary. A separate module would be just fine. We should have it in docs though.)
May I invite you to https://github.com/jotaijs ?

**Reply by iwoplaza** [reply]
I will make it a separate module then 🫡. Thanks for the quick response, and yes, I would love an invite to https://github.com/jotaijs 💜

**Reply by dai-shi** [reply]
Cool. Sent an invitation.

**Comment by sebinsua** [comment]
I quite frequently do these things within my own applications but haven't yet attempted to generalise and abstract this out in the way you have. I have `swr` wrappers (for getting the latest value) and many usages of `unwrap` and conditional checks, etc (to avoid aspects of my application needing loading indicators).

I agree that there should be a nicer way of doing these things, but some of this API I find kind of unusual. Like, calling a function `derive` when really any atom can "derive" -- it doesn't really describe to me what is different about what this does compared to just "deriving" from an async atom normally.

And then internally we have `soon` and `soonAll` but these are also not very descriptive of what they do either. I was thinking that there is some similarity between `soon` and `unwrap` -- `unwrap` always unwraps a value (and doesn't apply a transform), while this applies a transformation and then only unwraps to a synchronous value if the underlying atom does not contain a promise.

I don't have a good name for either right now though. It's kind of like a `preferSyncAtom` that takes a list of atoms it's dependent on and then tries to derive synchronously before falling back to an asynchronous derive?! On the other hand, the API is radically different from that of an atom, because you can't use `get` internally without running into the original issue you resolved, etc.

Or maybe another way of looking at it, is that it could be called something like `transformPreferUnwrap` or maybe `derivePreferUnwrap`...

**Reply by iwoplaza** [reply]
Thank you for the comment! Glad to see I am not the only one struggling with this. I agree that `derive` is very general, but I so far have not found a better name for it. My thinking on that was that I want to derive the inherent value of other atoms, irrespective of when they are available (known or a promise).

I do like the suggestions, maybe something with `unwrap`, `transform` or `prefer` would be better. I do worry that using the unwrap naming may clash and cause confusion with the already existing `unwrap` utility though.


```ts
const usersAtom: Atom<User[] | Promise<User[]>> = ...;
const nameFilterAtom = atom('');

// typeof filteredUsersAtom is Atom<User[] | Promise<User[]>>
const filteredUsersAtom = atomWithUnwrap(
  [usersAtom, nameFilterAtom],
  (users, nameFilter) => {
    if (!nameFilter) {
      return users;
    }
    
    return users.filter((user) => user.name.includes(nameFilter));
  }
);
```

**Comment by PhilipGrefe** [comment]
Hi all, I'd like to share my use case since I'm also struggling with potentially async values.

I'm writing a mobile app using react-native and [expo-sqlite](https://docs.expo.dev/versions/latest/sdk/sqlite/) as my database. Reads and writes are async. My atoms typically look like so:

```typescript
export const fooAtom = atomWithStorage(
  'foo',
  Promise.resolve(undefined as string | undefined), // trying to get the correct type inference here 🤔 🤷‍♂️
  {
    getItem: async () =>  await db.query.fooTable.findFirst() // some query,
    setItem: async (key, newFoo) => {
      await db.update() // some updater
    },
    removeItem: async () => {
      await db.delete() // some deleter
    },
    subscribe: // some subscriber with DB-listener
  },
  { getOnInit: true },
)
```

For now, I simply `unwrap` all usages of `atomWithStorage`. If I don't do that, the components disappear for a split-second (not using Suspense).

Looking forward to future updates here 🥳

**Reply by morganwu277** [reply]
maybe ... `Promise.resolve(undefined as string | undefined)` => `undefined as string | undefined` ?

**Comment by iwoplaza** [comment]
Based on the feedback, and my personal feelings since the initial proposal (released as [jotai-derive](https://github.com/jotaijs/jotai-derive)), I would like to reattempt a solution with better DX and more intuitive terminology.

> I agree that there should be a nicer way of doing these things, but some of this API I find kind of unusual. Like, calling a function derive when really any atom can "derive" -- it doesn't really describe to me what is different about what this does compared to just "deriving" from an async atom normally.

I agree that the "derive" API can be misleading, and making it different from the vanilla `atom((get) => ...)` API requires a shift in thinking and refactors to existing code from devs.

My proposal is to change the name of the project to `jotai-eager`, and provide an API that more closely mirrors the vanilla `atom` API: `eagerAtom`. "Eager" in this context means that we compute the atom's value synchronously if all of it's dependencies' values have been fulfilled (which is not the case for vanilla atoms).

```ts
import { eagerAtom } from 'jotai-eager';

const usersAtom: Atom<User[] | Promise<User[]>> = ...;
const nameFilterAtom = atom('');

// typeof filteredUsersAtom is Atom<User[] | Promise<User[]>>
const filteredUsersAtom = eagerAtom((get) => {
  const users = get.await(usersAtom);
  const nameFilter = get(nameFilterAtom);
  if (!nameFilter) {
    return users;
  }
  
  return users.filter((user) => user.name.includes(nameFilter));
});
```
When using `eagerAtom`, in addition to the `get` function that vanilla atoms can use, `get.await` either returns a fulfilled value, or throws to "suspend execution" (similar to Jotai v1 asynchronous atoms, React's `use`, etc.). Note that the thrown value will not reach outside of the atom's boundary, and will be handled internally.

This model greatly simplifies the amount of work needed to refactor a vanilla async atom into an `eagerAtom`, which boils down to changing `atom` to `eagerAtom`, removing `async` from the getter function, and swapping `get` and `await` keywords' positions.

It also makes conditional deriving very intuitive, where previously it was out of scope for the `derive` API, and required the use of the more complicated `soon` and `soonAll` APIs.

## Limitations/potential problems
- Discarding of work prior to `get.await`, and having to "restart" the atom computation once the Promise fulfills
- Having to reimplement `Promise.all`. (`get.await` could accept multiple arguments, or `get.awaitAll(...)` API)
- Complicating the use of `try {} catch {}` in eager atoms (can be disclosed in docs...?)

@dai-shi @sebinsua I would love to get your feedback on this proposal 🙏

**Reply by dai-shi** [reply]
A few instant comments:
- I generally like the idea to explore new API based on user feedback.
- `jotai-derive` would be heavily used by some users, so the migration path will be important.
- I don't particularly like the `get.await` api as it's not minifiable.


I would like to invite @uhyo for this discussion.

**Reply by iwoplaza** [reply]
Thanks for the quick response! 💜

> I don't particularly like the get.await api as it's not minifiable.

Does adding `get.await` make the `eagerAtom` api not minifiable in whole, or does having the `.await` method on `get` couple it together with `eagerAtom`? (meaning if `eagerAtom` is used, code for `get.await` is also included)

**Reply by iwoplaza** [reply]
Now that I think about it, since awaiting promise-ful atoms is the main use-case of eager atoms, maybe the default `get` should behave like `get.await` already?
```ts
const usersAtom: Atom<User[] | Promise<User[]>> = ...;
const nameFilterAtom = atom('');

const filteredUsersAtom = eagerAtom((get) => {
  const users = get(usersAtom); // => User[]
  const nameFilter = get(nameFilterAtom); // => string
  if (!nameFilter) {
    return users;
  }
  
  return users.filter((user) => user.name.includes(nameFilter));
});
```
I could also enforce on the type-level that the getter function passed into `eagerAtom` cannot return a Promise, which would nullify the benefits of using an eager atom.

**Reply by dai-shi** [reply]
> or does having the `.await` method on `get` couple it together with `eagerAtom`? (meaning if `eagerAtom` is used, code for `get.await` is also included)

Partially. `.await` will appear in user code too. It's just my preference not to use methods, and that's how I do it in Jotai core. But the final decision is on you.

**Reply by dai-shi** [reply]
> maybe the default `get` should behave like `get.await` already?

Sounds good to me.

**Reply by uhyo** [reply]
Hi! I'm someone who really likes the `soon` function and other related idea there 👋  I'm maybe a bit late but here's my thoughts as requested:

First of all I like the term “eager”. It represents the actual role of the current “derive” API very well. As someone who writes many articles to advertise ideas I like, names exactly matching their body is important.

Regarding the API, while I am fine with `get.await`, but making the default `get` behave like that would be good too. For the Suspense-like mechanism I have a mixed feeling:

- Since it is similar to Suspense from React, hopefully users are already familiar with its behavior and the try-catch footgun.
- This would require derived atom implementions to be “pure” like how React components should be so. Theoretically this has already been the case, but the requirement would be a harder one when `eagerAtom` is in practical use.

**Reply by iwoplaza** [reply]
@uhyo Thank you so much for the feedback! I am glad that the ideas 〰️ _resonate_ 〰️ with others.
> This would require derived atom implementions to be “pure” like how React components should be so. Theoretically this has already been the case, but the requirement would be a harder one when eagerAtom is in practical use.

Hmm... if we think of the current API, as in `derived` and `soon`/`soonAll`, the former was meant as a high-level solution for the most common use cases, whereas `soon`/`soonAll` were meant to fill the gap for more advanced uses.
If I keep the `soon` and `soonAll` APIs, they can still fulfill that role for cases where purity of atoms is questionable, while improving the DX for the more common use cases.

**Reply by sebinsua** [reply]
Hi @iwoplaza I see you've updated `jotai-eager` and I like that you've introduced an `eagerAtom` that uses `throw` to make async atoms behave as if they can be read synchronously as this allows the abstraction to completely hide the micro-suspense problem.

I was [attempting a similar thing here](https://github.com/pmndrs/jotai/discussions/3044) but (as part of a `withSwr` utility) and I ran into issues when composing multiple atoms that did this because (I think) of how I tried to let React handle the recursive suspending. To be honest, I suspect that what you've created could help me to finish!

How does `resolveSuspension` work?

**Reply by iwoplaza** [reply]
> Hi @iwoplaza I see you've updated jotai-eager and I like that you've introduced an eagerAtom that uses throw to make async atoms behave as if they can be read synchronously as this allows the abstraction to completely hide the micro-suspense problem.

Happy to hear it! This makes me more confident that this is the right move 🚀

> How does resolveSuspension work?

I used to depend on Jotai attaching a status to promises (an extra property with a status: 'fulfilled' | 'rejected' | 'pending', the fulfilled value if it exists, and a reject reason if it exists). This was recently changed in Jotai internals, and is no longer going to be handled there. I am now tracking each Promise that goes through my APIs myself, and `resolveSuspension(compute)` works like the following:
- Run `compute` inside of a try and catch, and return it's result.
- If caught an error, and it's EagerError, extract the Promise that caused the suspension and chain it to run `resolveSuspension(compute)` recursively when the value of the promise is available.

Hope that answers it, feel free to follow up if you need more context 🙏

**Comment by iwoplaza** [comment]
@dai-shi Hi! 👋 
I was doing maintenance to `jotai-derive` and bumping dependencies, and I noticed that some of the tests are breaking in the latest Jotai release. I am not sure when this change was made (I assume sometime after React 19), but attaching Promise statuses is now the responsibility of `useAtomValue`. If I recall correctly, this was previously the responsibility of the store, meaning that `jotai/vanilla` users could leverage the same benefits of `jotai-derive` as React users could.

**Reply by dai-shi** [reply]
👉 #3069
